internal/serve/httphandler/wallet_creation_handler.go

```
@@ -0,0 +1,119 @@
+	}
+
+	if validator.HasErrors() {
+		return httperror.BadRequest("", nil, validator.Errors)
+	}
+
+	return nil
+}
+
+type WalletResponse struct {
+	ContractAddress string                    `json:"contract_address,omitempty"`
+	Status          data.EmbeddedWalletStatus `json:"status"`
+}
+
+func (h WalletCreationHandler) getTenantFromContext(ctx http.Request) (*tenant.Tenant, *httperror.HTTPError) {
+	currentTenant, err := tenant.GetTenantFromContext(ctx.Context())
+	if err != nil || currentTenant == nil {
+		return nil, httperror.InternalError(ctx.Context(), "Failed to load tenant from context", err, nil)
+	}
+	return currentTenant, nil
+}
+
+func (h WalletCreationHandler) CreateWallet(rw http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+
+	var reqBody CreateWalletRequest
+	if err := httpdecode.DecodeJSON(req, &reqBody); err != nil {
+		httperror.BadRequest("", err, nil).Render(rw)
+		return
+	}
+
+	if err := reqBody.Validate(); err != nil {
+		err.Render(rw)
+		return
+	}
+
+	currentTenant, httpErr := h.getTenantFromContext(*req)
+	if httpErr != nil {
+		httpErr.Render(rw)
+		return
+	}
+
+	err := h.embeddedWalletService.CreateWallet(ctx, currentTenant.ID, reqBody.Token, reqBody.PublicKey)
+	if err != nil {
+		switch err {
+		case services.ErrCreateWalletInvalidToken:
+			httperror.BadRequest("Invalid token", err, nil).Render(rw)
+		case services.ErrCreateWalletInvalidStatus:
+			httperror.BadRequest("Wallet status is not pending for token", err, nil).Render(rw)
+		default:
+			httperror.InternalError(ctx, "Failed to create wallet", err, nil).Render(rw)
+		}
+		return
+	}
+
+	resp := WalletResponse{
+		Status: data.PendingWalletStatus,
+	}
+	httpjson.RenderStatus(rw, http.StatusOK, resp, httpjson.JSON)
+}
+
+func (h WalletCreationHandler) GetWallet(rw http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+	token := strings.TrimSpace(req.URL.Query().Get("token"))
```

Do you want to error out with a 400 is the `token` is missing?

---

Nice catch, this is missing.